### ISSUE TITLE
Please fix OSSFuzz build failure with libjpeg-turbo

### ISSUE THREAD

**Issue by dcommander** [google/oss-fuzz#4931]
I keep getting spammed about this:
https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=27916
but I have no ability to fix it, since the problem is on your end.  It's probably due to having a too-old version of NASM on your build machine.  I'm sick of receiving automatic reminders to fix it, when you haven't granted me any authority to do so.

**Comment by jonathanmetzman** [comment]
I think there is a misunderstanding here because libjpeg-turbo is in a somewhat unique position.
Unlike most projects that were integrated into OSS-Fuzz, the integration of libjpeg-turbo was done without the involvement of the primary maintainer (you) so you might be unfamiliar with how these integrations are maintained.

It's generally the responsibility of OSS-Fuzz projects to ensure that they can build their latest code in our docker image.
The exception to this is (in some cases) when we (the OSS-Fuzz maintainers) change the image causing projects to break. I don't think this is one of those cases.

You or the person who integrated this project should modify the integration (e.g. by downloading the required version of NASM) so that the project can build in OSS-Fuzz.
Let me know if you need guidance here.
Sorry for the confusion.

**Comment by dcommander** [comment]
I did not integrate libjpeg-turbo into OSS-Fuzz, so I have no knowledge of how to fix this issue, nor do I have any funding to work on such things right now.  libjpeg-turbo incurred a serious funding deficit in order to finish the 2.1 beta release (which, notably, had several major features that are of direct benefit to Android and Chrome), and until that deficit is addressed, I have no ability to work on anything except the most high-priority bug fixes.  At our current rate of funding, it will take more than a year for the deficit to be addressed.

Long story short: someone else will have to fix this.

**Comment by inferno-chromium** [comment]
@dcommander - would you be interested in getting funding through OSS-Fuzz rewards, libjpeg-turbo is an important project to OSS Community and should qualify for atleast $10K+ funding - https://google.github.io/oss-fuzz/getting-started/integration-rewards/ and https://google.github.io/oss-fuzz/advanced-topics/ideal-integration/

**Comment by dcommander** [comment]
Yes, very interested.  In addition to helping our project's funding woes, I could also take advantage of the opportunity to move our CI builds to GitHub Actions and thus get them working again (Travis CI unfortunately wants to charge open source developers money to continue using their systems.)

**Comment by inferno-chromium** [comment]
@dcommander - so those two links should tell on what things to do and then the form to fill to get the reward. Straightforward process, let us know if you have any questions.

**Comment by inferno-chromium** [comment]
Closing this here then.

**Comment by dcommander** [comment]
It may seem like a "straightforward" process to you, as someone who has intimate familiarity with the system, but after spending hours trying to figure it out, I am no closer.  The documentation doesn't make it clear how the various pieces connect.  I understand that I need to build a fuzzer target with libjpeg-turbo, but what if I need multiple fuzzer targets to cover multiple API entry points?  (The existing target only covers one decompression code path.)  How does OSS-Fuzz invoke those targets?  Where are the existing seed corpus files for libjpeg-turbo?  Where and how would those be integrated into my repository?

Please understand that our project currently has a 130-hour funding deficit, so there's a limit to how much additional pro bono work I can do in order to unlock this additional funding from Google.

**Comment by inferno-chromium** [comment]
Hello @dcommander, did you get a chance to look at these files for libjpeg_turbo project - https://github.com/google/oss-fuzz/tree/master/projects/libjpeg-turbo. If you read those files and this doc - https://google.github.io/oss-fuzz/getting-started/new-project-guide/, that should help you understand current integration. I am sorry i thought you understood the process and that is why i pointed towards https://google.github.io/oss-fuzz/advanced-topics/ideal-integration/ (which comes after understanding initial integration).

Fuzz target - only one - https://github.com/google/oss-fuzz/blob/master/projects/libjpeg-turbo/libjpeg_turbo_fuzzer.cc
You can add more fuzz targets to target more interesting apis.
These fuzz target need to be specified in this build file - https://github.com/google/oss-fuzz/blob/master/projects/libjpeg-turbo/build.sh#L23 for building and put in $OUT
Seed corpus is provided here - https://github.com/google/oss-fuzz/blob/master/projects/libjpeg-turbo/Dockerfile#L23 and https://github.com/google/oss-fuzz/blob/master/projects/libjpeg-turbo/build.sh#L26
DockerFile - specifies dependencies to install and place to get code, shouldn't need more changes - https://github.com/google/oss-fuzz/blob/master/projects/libjpeg-turbo/Dockerfile

By integrating in repository, can start moving many of those files in libjpeg-turbo repo itself, see https://google.github.io/oss-fuzz/advanced-topics/ideal-integration

**Comment by dcommander** [comment]
@inferno-chromium Since it appears as if project funding has finally been secured, I had a follow-up question about OSS-Fuzz integration.  I understand the basics of how to move the fuzzer target out of the OSS-Fuzz repository and into the libjpeg-turbo repository, but one thing I don't understand is where the existing corpus files came from.  They appear to be hosted on an external site (https://lcamtuf.coredump.cx).

**Comment by inferno-chromium** [comment]
> @inferno-chromium Since it appears as if project funding has finally been secured, I had a follow-up question about OSS-Fuzz integration. I understand the basics of how to move the fuzzer target out of the OSS-Fuzz repository and into the libjpeg-turbo repository, but one thing I don't understand is where the existing corpus files came from. They appear to be hosted on an external site (https://lcamtuf.coredump.cx).

lcamtuf@ worked at google before, and these corpus files came from AFL fuzzer written by him - https://github.com/google/AFL/tree/master/testcases/images

**Comment by dcommander** [comment]
@inferno-chromium The AFL fuzzer source only contains a single JPEG image.  Were the rest of the images in the tarball automatically generated?  If not, then where did they come from?  I'm mainly trying to figure out whether it makes sense to maintain my own seed corpus repository.  If it does, then I would like to incorporate the existing images in that tarball, but I would need to know which license applies to them.

**Comment by inferno-chromium** [comment]
> @inferno-chromium The AFL fuzzer source only contains a single JPEG image. Were the rest of the images in the tarball automatically generated? If not, then where did they come from? I'm mainly trying to figure out whether it makes sense to maintain my own seed corpus repository. If it does, then I would like to incorporate the existing images in that tarball, but I would need to know which license applies to them.

It might be mix of autogenerated and starting with an existing image corpus, i dont know for sure. but license is all apache 2.0 - https://github.com/google/AFL/blob/master/LICENSE

**Comment by dcommander** [comment]
Sorry to bother you again, but at this point, I've spent in excess of 50 hours on OSS-Fuzz integration.  I need assistance from someone who is more familiar with the system than I am.  Specifically, there are several crashes that I've fixed and selected "Redo Task" on, but they are still showing up as open crashes.  I need some insight as to when OSS-Fuzz will retest those crashes and close them.  As you can understand, I'm now using OSS-Fuzz to determine whether there are any critical bugs left in libjpeg-turbo before it leaves beta, and the inability to clear these crashes in a timely manner is holding up the libjpeg-turbo 2.1 release.

**Comment by inferno-chromium** [comment]
OSS-Fuzz builds are generated once a day around 10-12pm PST. and clusterfuzz automatically tests for fixes once a day around 6-8am PST. you can force redo, but it will only declare fixed if it has a new build. if you want more frequent builds a day, we allow upto 4, see https://google.github.io/oss-fuzz/getting-started/new-project-guide/#build_frequency

among the 3 bugs left open, https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32964 is fixed [where you clicked redo], clusterfuzz should close that bug soon, you can see it fixed in metadata on page below
```
[2021-04-07 14:30:41 UTC] oss-fuzz-linux-zone6-host-g6gl-6: Progression task started: r202104070621.
[2021-04-07 14:30:45 UTC] oss-fuzz-linux-zone6-host-g6gl-6: Progression task finished: fixed in range r202104060609:202104070621.
```
remaining 2 bugs open - https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32972&q=libjpeg-turbo, https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32982&q=libjpeg-turbo
you can wait a day to see if new builds fix it.

**Comment by dcommander** [comment]
I've already waited more than a day for the timeouts to clear.  Those issues were fixed on Monday.

**Comment by inferno-chromium** [comment]
> I've already waited more than a day for the timeouts to clear. Those issues were fixed on Monday.

Timeouts are all fixed and closed, those are other issues, can you click the links i pasted in last comment to see. Also, ClusterFuzz cron just closed https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=32964#c3 as expected. Automation can take upto a day.

**Comment by dcommander** [comment]
No, they are not all fixed and closed.
![image](https://user-images.githubusercontent.com/651941/113897100-2ebaa200-9790-11eb-9237-8e4ffa52b120.png)

**Comment by dcommander** [comment]
And yes, I'm looking into the other two open issues.  I'm mainly concerned with the timeouts right now.  There was a third timeout that automatically cleared, so I don't know why the other two didn't.

**Comment by inferno-chromium** [comment]
> And yes, I'm looking into the other two open issues. I'm mainly concerned with the timeouts right now. There was a third timeout that automatically cleared, so I don't know why the other two didn't.

Ignore the unreproducible ones [see reliably reproduces is greyed out]. Completely ignore them. This is why those bugs are never filed in monorail issue tracker system. that web ui is just an easy way to see things. unreproducibles ones are only filed a bug if we see them consistently for 14 days. this just shows we saw them on Apr 3 and Apr4, but not after that. If you really want a real-time look at things, try https://oss-fuzz.com/crash-stats that shows which crashes are still triggering.

**Comment by dcommander** [comment]
But the timeouts were valid.  They were an example of a known issue with the progressive JPEG format (specifically, LJT-01-003 described [here](https://libjpeg-turbo.org/pmwiki/uploads/About/TwoIssueswiththeJPEGStandard.pdf).)  I worked around the timeouts in https://github.com/libjpeg-turbo/libjpeg-turbo/commit/c81e91e8ca34f4e8b43cf48277c2becf3fe9447d by introducing a TurboJPEG API flag that limits the number of progressive JPEG scans.  (That flag is also generally useful for any application using the TurboJPEG API that wishes to similarly ignore progressive JPEG images with an unreasonably high number of scans.)

I'm just trying to get a handle on best practices and expectations.  If memory serves, none of the timeouts was flagged as reliably reproducible, even though they were valid.  If my memory on that is correct, then I don't understand why the first timeout was assigned an issue and then automatically cleared whereas the other two timeouts weren't.

**Comment by inferno-chromium** [comment]
check this out - https://oss-fuzz.com/crash-stats?block=hour&days=3&group=platform&number=count&project=libjpeg-turbo&q=timeout&sort=total_count
this is a great way to see if timeout are happening at all now, looks like nothing is happening now, no unreproducible or reproducible ones, so you did fix something :)

**Comment by dcommander** [comment]
Great.  So now could you please answer my direct questions above:
- Why was only one timeout assigned an issue and then automatically cleared but not the other two?
- When should I expect the other two to automatically clear?

**Comment by inferno-chromium** [comment]
> Great. So now could you please answer my direct questions above:
> * Why was only one timeout assigned an issue and then automatically cleared but not the other two?

That timeout was reproducible, other two were not. timeouts can be flaky, so even if it was reproducible on one revision, it became flaky on next and then got closed.

> * When should I expect the other two to automatically clear?

unreproducibles get deleted after 7 days if they are no longer seen.

**Comment by dcommander** [comment]
The only issue left is an MSan false positive with `fread()`, and I have no idea how to work around it.  (`__msan_unpoison()` doesn't seem to work.)  Please advise.  Unfortunately, there is no way to avoid C library file I/O, because that's the only way to feed data into the BMP and PPM readers.
